protocol/consistencychecks.go

```diff
@@ -222,43 +241,53 @@ func (cc *ConsistencyChecks) verifyReturnedPromise(tb *m.TemporaryBinding,
 			return ErrorBadIndex
 		}
 
-		cc.TBs = append(cc.TBs, tb)
+		// verify TB's value
+		if !bytes.Equal(tb.Value, key) {
+			return ErrorBadPromise
+		}
+
+		cc.TBs[uname] = tb
 	}
 
 	return nil
 }
 
 // verifyFulfilledPromise verifies issued TBs was inserted
 // in the directory as promised.
-func (cc *ConsistencyChecks) verifyFulfilledPromise(str *m.SignedTreeRoot,
+func (cc *ConsistencyChecks) verifyFulfilledPromise(uname string,
+	str *m.SignedTreeRoot,
 	ap *m.AuthenticationPath) error {
-	if cc.isUseTBs {
-		backed := cc.TBs[:0]
-		for _, tb := range cc.TBs {
-			if cc.CurrentEpoch+1 == str.Epoch {
+	if cc.useTBs {
+		for name, tb := range cc.TBs {
+			if cc.CurrentEpoch+1 == str.Epoch &&
+				name == uname {
 				if !tb.Verify(ap) {
 					return ErrorBrokenPromise
 				}
-			} else if cc.CurrentEpoch == str.Epoch {
-				// keep current epoch's returned promises
-				// for future verifications
-				backed = append(backed, tb)
+				delete(cc.TBs, uname)
+			} else if cc.CurrentEpoch < str.Epoch {
+				// clear all issued promises since they have been verified
+				// or the client has missed some epochs
```

Comment:
I think what I'm suggestion here is slightly different than the other discussion. What was going on before is that we were verifying every TB for the current epoch for every single AP that the client was receiving, regardless of the user/index the TB was for. What I'm suggesting now is that the client verify old TBs only for its own user during monitoring in case it missed any epochs. For example, let's say that the client received a TB for its user alice at epoch `t=2` and it went offline immediately after. When the client comes back online at `t=5` it should ensure that alice's old TB was properly included at `t=3` while also verifying the other missed epochs. I think in the current implementation, the TB from `t=2` would be deleted, no?